sound/soc/sof/intel/hda.c

```diff
@@ -220,12 +220,11 @@ static int hda_sdw_exit(struct snd_sof_dev *sdev)
 
 	hdev = sdev->pdata->hw_pdata;
 
-	hda_sdw_int_enable(sdev, false);
-
 	if (hdev->sdw)
 		sdw_intel_exit(hdev->sdw);
 	hdev->sdw = NULL;
 
+	hda_sdw_int_enable(sdev, false);
```

Comment:
can we give an example of when "Codecs may need to access sdw registers via sdw bus in the exit process".

I don't fully buy the argument. IIRC the initial code was intentional so that we wouldn't have ANY interrupts while removing the devices.

Reply:
> can we give an example of when "Codecs may need to access sdw registers via sdw bus in the exit process".

When we run `sudo modprobe -r snd_sof_pci_intel_lnl`, `snd_sof_runtime_resume` will be trigger first. And `hda_sdw_process_wakeen` will be called when dsp boot. Due to `intel_shim_check_wake` return 1 unconditionally, `soundwire_intel.link`s will be powered up. And the codecs will be attached. 
SCP Msg trf timed out will happen if `hda_sdw_exit` is called before codec is initialized, and SDW interrupt is disabled.

Please see the log in detail.
[lnl_remove.txt](https://github.com/thesofproject/linux/files/14381095/lnl_remove.txt)

> I don't fully buy the argument. IIRC the initial code was intentional so that we wouldn't have ANY interrupts while removing the devices.

I agree that we don't want any interrupts while removing the devices. However, the interrupts are also used in sdw commend transfer. Can we cancel the sdw commend transfer before disabling SDW interrupt?

Reply:
@plbossart Alternative solution #4838. Instead of canceling the sdw commend, #4838 wait for them be finished.